contracts/HomoraIBSwap.sol

```diff
@@ -0,0 +1,220 @@
+    function ibToToken(address ibToken, uint256 ibTokenAmount)
+        external
+        view
+        returns (uint256)
+    {
+        CYToken cyToken = CYToken(SafeBox(ibToken).cToken());
+        uint256 tokenAmount =
+            ibTokenAmount.mul(cyToken.exchangeRateStored()).div(1e18);
+        return tokenAmount;
+    }
+
+    /// @notice convert an amount of token to that of the associated ibToken
+    /// @param ibToken the ibToken to get the token amount of
+    /// @param tokenAmount the amount of token to use for conversion
+    function tokenToIB(address ibToken, uint256 tokenAmount)
+        external
+        view
+        returns (uint256)
+    {
+        CYToken cyToken = CYToken(SafeBox(ibToken).cToken());
+        uint256 ibTokenAmount =
+            tokenAmount.mul(1e18).div(cyToken.exchangeRateStored());
+        return ibTokenAmount;
+    }
+
+    /// @notice swap an ibToken to another ibToken
+    /// @param tokenIn the input ibToken
+    /// @param tokenOut the desired output ibToken
+    /// @param amountIn the amount of tokenIn to swap
+    /// @param amountOutMin minimum amount of output tokens that must be received
+    /// for the transaction not to revert
+    /// @param deadline timestamp after which the transaction will revert
+    function swap(
+        address tokenIn,
+        address tokenOut,
+        uint256 amountIn,
+        uint256 amountOutMin,
+        uint256 deadline
+    ) external returns (uint256) {
+        require(tokenIn != tokenOut, "token-in-out-identical");
+        require(isIBToken[tokenIn], "token-in-not-supported");
+        require(isIBToken[tokenOut], "token-out-not-supported");
+
+        SafeBox safeboxIn = SafeBox(tokenIn);
+        safeboxIn.transferFrom(msg.sender, address(this), amountIn);
+        safeboxIn.withdraw(amountIn);
+
+        uint256 underlyingBalance;
+        address[] memory path = getPath(tokenIn, tokenOut);
+        uint256 outputAmount;
+
+        if (tokenIn != IBETHV2) {
+            IERC20 underlying = IERC20(safeboxIn.uToken());
+            underlyingBalance = underlying.balanceOf(address(this));
+        } else {
+            underlyingBalance = address(this).balance;
+        }
+
+        if (tokenIn != IBETHV2 && tokenOut != IBETHV2) {
+            uniswapRouter.swapExactTokensForTokens(
+                underlyingBalance,
+                0,
+                path,
+                address(this),
+                deadline
+            );
+        } else if (tokenIn == IBETHV2) {
+            uniswapRouter.swapExactETHForTokens{value: underlyingBalance}(
+                0,
+                path,
+                address(this),
+                deadline
+            );
+        } else if (tokenOut == IBETHV2) {
+            uniswapRouter.swapExactTokensForETH(
+                underlyingBalance,
+                0,
+                path,
+                address(this),
+                deadline
+            );
+        }
```

Comment:
Perhaps a better way to do this is to work with WETH when swapping, so you only need `swapExactTokensForTokens`. Then you just need to check if `tokenIn` is ETH related --> Then withdraw ETH -> wrap to WETH.

If `tokenOut` is ETH related --> unwrap WETH -> deposit to ibETH.